# MATH 210 Introduction to Mathematical Computing

## January 19, 2022

* Sequences: lists, tuples and range objects
* List comprehensions
* Builtin Python functions: `type`, `print`, `sum`

## Sequences

There are 3 main sequence types: lists, tuples and range objects.

We use square brackets `[ ... ]` to define a Python `list`:

In [1]:
squares = [1,4,9,16,25,36,49,64,81,100]

In [2]:
squares

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]

In [3]:
type(squares)

list

We use parentheses `( ... )` to define a Python `tuple`:

In [4]:
date = (2022,1,19)

In [5]:
date

(2022, 1, 19)

In [6]:
type(date)

tuple

The differences between a Python `list` and a `tuple` are technical and not really important to us. If you want to look it up, a `list` is mutable (we can modify entries) and `tuple` is not mutable (cannot modify entries).

For example, let's modify the first entry (at index 0) of our list `squares`:

In [7]:
squares[0]

1

In [8]:
squares[0] = -1

In [9]:
squares

[-1, 4, 9, 16, 25, 36, 49, 64, 81, 100]

But we get an error if we try to change the tuple `date`:

In [10]:
date[0]

2022

In [11]:
date[0] = 1922

TypeError: 'tuple' object does not support item assignment

A `range` object is an efficient sequence generator. A range object does not have any entries until you ask for them. The command `range(a,b)` creates a range object which begins at integer `a` and increments by 1 up to but not including integer `b`. Notice that the length of the sequence is `b - a`.

In [12]:
numbers = range(1,11)

In [13]:
numbers

range(1, 11)

In [14]:
type(numbers)

range

Use the Python function `list` to converts a `range` to a `list`.

In [15]:
list(numbers)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

The command `range(a,b,n)` creates a range object which begins at integer `a` and increments by `n` up to but not including integer `b`.

In [16]:
evens = range(0,21,2)

In [17]:
list(evens)

[0, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20]

## List Comprehensions

A list comprehension is a simple construction to create lists. The syntax of a list comprehension is:

```
[expression for item in iterable]
```

where:
* `expression` is a Python expression like `x**2`
* `item` is a variable name like `x` as in the example in line above
* `iterable` is any sequence of values

In [18]:
squares = [n**2 for n in range(1,17)]

In [19]:
print(squares)

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144, 169, 196, 225, 256]


Can we use a list comprehension to generate the sequence of Fibonacci numbers? Recall, Fibonacci numbers are $1,1,2,3,5,8,13,21,...$, given by the recursive formula

$$
x_n = x_{n-1} + x_{n-2}
$$

No! Lists comprehensions only work for explicit sequences where we have a expression for each term as a function of the index `n` only.

## Builtin Functions

Python has many builtin functions. We have seen a few already like `type`, `print`, `list`, and `range`. There is also `sum` which adds all the entries in a sequence.

In [20]:
sum([1,2,3,4,5])

15

Let's do an example. The Taylor series of $\ln(1+x)$ is

$$
\ln(1+x) = \sum_{n = 1}^{\infty} (-1)^{n+1} \frac{x^n}{n}
$$

Let's use a partial sum to approximate $\ln(1.5)$ ($x=0.5$)

$$
\ln(1.5) \approx \sum_{n = 1}^{N} (-1)^{n+1} \frac{(0.5)^n}{n}
$$

In [21]:
N = 25
terms = [(-1)**(n+1)*(0.5)**n/n for n in range(1,N+1)]
result = sum(terms)
print(result)

0.40546510849507883


Use a graphing calculator to find the first 9 decimal values of the exact value

$$
\ln(1.5) \approx 0.405465108
$$